lib/avromatic/model/validation.rb

```diff
@@ -2,52 +2,71 @@ module Avromatic
   module Model
     module Validation
       extend ActiveSupport::Concern
+      include ActiveModel::Validations
 
-      module ClassMethods
-
-        # Returns an array of messages
-        def validate_nested_value(value)
-          case value
-          when Avromatic::Model::Attributes
-            validate_record_value(value)
-          when Array
-            value.flat_map.with_index do |element, index|
-              validate_nested_value(element).map do |message|
-                "[#{index}]#{message}"
-              end
-            end
-          when Hash
-            value.flat_map do |key, map_value|
-              # keys for the Avro map type are always strings and do not require
-              # validation
-              validate_nested_value(map_value).map do |message|
-                "['#{key}']#{message}"
-              end
-            end
-          else
-            []
+      EMPTY_ARRAY = [].freeze
+
+      def self.missing_nested_attributes(attribute, value)
+        if value.is_a?(Array)
+          results = []
+          value.each_with_index do |element, index|
+            nested_results = missing_nested_attributes("#{attribute}[#{index}]", element)
+            results.concat(nested_results)
+          end
+          results
+        elsif value.is_a?(Hash)
+          results = []
+          value.each do |key, element|
+            nested_results = missing_nested_attributes("#{attribute}['#{key}']", element)
+            results.concat(nested_results)
+          end
+          results
+        elsif value.is_a?(Avromatic::Model::Validation)
```

Comment:
I think the comment I made was that this condition was surprising/potentially confusing. I assume that this is checking for an Avromatic model. Is this a condition that is used elsewhere in the code? Or is there something that would be more intention revealing?

Reply:
This condition doesn't appear anywhere else. Do you think it would be cleaner to check `respond_to?(:missing_avro_attributes)`?

Reply:
I think the condition that I've used for similar purposes elsewhere is `value.is_a?(Avromatic::Model::Attributes)`. I don't know if that is any better

Another option may be to create a helper method here, `model_instance?(value)`, and then define `#model_instance?` using the existing expression or one of the other suggestions here.

Reply:
I agree that `value.is_a?(Avromatic::Model::Attributes)` isn't much clearer :) I think I'll just use `value.respond_to?(:missing_avro_attributes)` since that accurately captures the contract that this branch of the conditional actually cares about.